src/backend/access/appendonly/appendonlywriter.c

```diff
@@ -778,6 +778,54 @@ RegisterSegnoForCompactionDrop(Oid relid, List *compactedSegmentFileList)
 	return;
 }
 
+List *
+UpdateSegnoAfterCompaction(Oid relid, List *compactedSegmentFileList)
+{
+	AORelHashEntryData *aoentry;
+	Relation	aosegrel;
+	bool	   *awaiting_drop = NULL;
+	int			i;
+
+	Assert(Gp_role != GP_ROLE_EXECUTE);
+	if (Gp_role == GP_ROLE_UTILITY)
+	{
+		return compactedSegmentFileList;
+	}
+
+	if (compactedSegmentFileList == NIL)
+	{
+		return compactedSegmentFileList;
+	}
+
+	aosegrel = relation_open(relid, AccessShareLock);
+	awaiting_drop = get_awaiting_drop_status_from_segments(aosegrel);
```

Comment:
There are dispatch results for each vacuum operation. Did you research possibility to use/extend them?

Reply:
> There are dispatch results for each vacuum operation. Did you research possibility to use/extend them?

Not yet, but it seems that information about the state of the segment file can be transmitted through them.

Reply:
I tried to do this, but it turned out somehow too monstrous: https://github.com/arenadata/gpdb/pull/549

Reply:
@RekGRpth, what about rewriting GetTotalTupleCountFromSegments()? 
Its call stack is:

```
GetTotalTupleCountFromSegments() at appendonlywriter.c:1375
UpdateMasterAosegTotalsFromSegments() at appendonlywriter.c:1626
vacuum_rel () at vacuum.c:2579
vacuum_rel_ao_phase () at vacuum.c:2252
vacuumStatement_Relation () at vacuum.c:890
vacuum () at vacuum.c:353
```

The GetTotalTupleCountFromSegments() function executes "`SELECT tupcount, segno FROM pg_aoseg.pg_aoseg_16384`" and
the get_awaiting_drop_status_from_segments() function executes "`SELECT state, segno FROM pg_aoseg.pg_aoseg_16384`".
Can we execute one query "`SELECT tupcount, state, segno FROM pg_aoseg.pg_aoseg_16384`" instead of two queries?

Reply:
> what about rewriting GetTotalTupleCountFromSegments()? Its call stack is:
> 
> ```
> GetTotalTupleCountFromSegments() at appendonlywriter.c:1375
> UpdateMasterAosegTotalsFromSegments() at appendonlywriter.c:1626
> vacuum_rel () at vacuum.c:2579
> vacuum_rel_ao_phase () at vacuum.c:2252
> vacuumStatement_Relation () at vacuum.c:890
> vacuum () at vacuum.c:353
> ```
> 
> The GetTotalTupleCountFromSegments() function executes "`SELECT tupcount, segno FROM pg_aoseg.pg_aoseg_16384`" and the get_awaiting_drop_status_from_segments() function executes "`SELECT state, segno FROM pg_aoseg.pg_aoseg_16384`". Can we execute one query "`SELECT tupcount, state, segno FROM pg_aoseg.pg_aoseg_16384`" instead of two queries?

Yes, you are right, this also works, and it looks more optimal. See https://github.com/arenadata/gpdb/pull/562

Reply:
done